File: src/io/gitignore.ts
import { join } from "node:path";
import { writeFileAtomic } from "./write.ts";

const BEGIN = "# menv (managed block)";
const END = "# end menv";

// Idempotently maintain menv's block in .gitignore: entries are unioned (set
// semantics, original order kept), user lines outside the block are untouched.
export async function upsertManagedBlock(root: string, entries: string[]): Promise<void> {
  const file = Bun.file(join(root, ".gitignore"));
  const text = (await file.exists()) ? await file.text() : "";
  const lines = text.split("\n");
  const begin = lines.indexOf(BEGIN);
  const end = begin === -1 ? -1 : lines.indexOf(END, begin);

  const existing = begin !== -1 && end !== -1 ? lines.slice(begin + 1, end) : [];
  const merged = [...existing];
  for (const e of entries) if (!merged.includes(e)) merged.push(e);
  const block = [BEGIN, ...merged, END];

  let out: string[];
  if (begin !== -1 && end !== -1) {
    out = [...lines.slice(0, begin), ...block, ...lines.slice(end + 1)];
  } else {
    const head = text === "" ? [] : [...lines.slice(0, lines.at(-1) === "" ? -1 : lines.length), ""];
    out = [...head, ...block, ""];
  }
  await writeFileAtomic(root, ".gitignore", out.join("\n"));
}

File: go/internal/io/gitignore.go
package io

import (
	"os"
	"path/filepath"
	"strings"
)

const (
	gitignoreBegin = "# menv (managed block)"
	gitignoreEnd   = "# end menv"
)

// UpsertManagedBlock idempotently maintains menv's block in .gitignore.
// Entries are unioned (set semantics, original order kept); lines outside
// the block are left untouched.
func UpsertManagedBlock(root string, entries []string) error {
	path := filepath.Join(root, ".gitignore")
	data, err := os.ReadFile(path)
	text := ""
	if err == nil {
		text = string(data)
	} else if !os.IsNotExist(err) {
		return err
	}

	lines := strings.Split(text, "\n")
	begin := indexOf(lines, gitignoreBegin)
	end := -1
	if begin != -1 {
		end = indexOfFrom(lines, gitignoreEnd, begin)
	}

	existing := []string{}
	if begin != -1 && end != -1 {
		existing = lines[begin+1 : end]
	}
	merged := append([]string{}, existing...)
	for _, e := range entries {
		if !contains(merged, e) {
			merged = append(merged, e)
		}
	}
	block := append([]string{gitignoreBegin}, append(merged, gitignoreEnd)...)

	var out []string
	if begin != -1 && end != -1 {
		out = append(lines[:begin:begin], append(block, lines[end+1:]...)...)
	} else {
		// Append after existing content (with a blank separator if needed).
		head := lines
		if text == "" {
			head = nil
		} else {
			// Trim trailing empty line to avoid double-blank.
			if len(head) > 0 && head[len(head)-1] == "" {
				head = head[:len(head)-1]
			}
			head = append(head, "") // blank separator
		}
		out = append(head, append(block, "")...)
	}
	return WriteFileAtomic(root, ".gitignore", []byte(strings.Join(out, "\n")))
}

func indexOf(lines []string, s string) int {
	for i, l := range lines {
		if l == s {
			return i
		}
	}
	return -1
}

func indexOfFrom(lines []string, s string, from int) int {
	for i := from; i < len(lines); i++ {
		if lines[i] == s {
			return i
		}
	}
	return -1
}

func contains(lines []string, s string) bool {
	for _, l := range lines {
		if l == s {
			return true
		}
	}
	return false
}
